doc/developer/design/20210406_mapfilterproject.md

```diff
@@ -0,0 +1,155 @@
+Arbitrary sequences of these operations can always be reduced to one `Map`, then one `Filter`, then one `Project`.
+
+The proposal is to assemble and communicate this information (caveat: some modifications will be made) to sources:
+```rust
+pub struct MapFilterProject {
+    /// A sequence of expressions that should be appended to the row.
+    pub expressions: Vec<MirScalarExpr>,
+    /// Expressions that must evaluate to `Datum::True` for the output
+    /// row to be produced.
+    pub predicates: Vec<MirScalarExpr>,
+    /// A sequence of column identifiers whose data form the output row.
+    pub projection: Vec<usize>,
+}
+```
+The three members correspond exactly to one `Map`, then one `Filter`, then one `Project`.
+
+On instantiation each source would be supplied with a `&mut MapFilterProject` argument.
+Depending on the capabilities of the source, it could take any or all of the operations, and leave behind any work undone.
+A simple default would be to leave the argument untouched, communicating that it applied none of the operations.
+A thorough source could take the entire instance and leave behind an identity operator, communicating that there is no remaining work to be done.
+
+### Manipulation
+
+The `MapFilterProject` structure has several methods that are meant to facilitate its manipulation.
+For example, the `demand(&self)` method indicates which input columns will be examined by the instance.
+Additionally, the `permute(&mut self, ..)` method re-writes all column references as if the input had been permuted.
+These two together allow sources to determine which of their announced columns need to be constructed at all, and to rewrite a `MapFilterProject` instance to reference only those columns that were constructed.
+
+As an example, the `csv.rs` source provides the appearance of multiple columns of text.
+A `MapFilterProject` instance may only consult three of those columns, and ignore all of the others.
+Rather than form `Datum::String` instances for each input column, the source can only produce the three columns of interest and apply the re-written `MapFilterProject` directly to that smaller input.
+This is especially valuable in situations where input columns may be more complicated, for example Avro metadata, which can involve expensive decoding to prepare, but which can be entirely skipped if none of the contents are referenced.
+
+### Planning for evaluation
+
+The `MapFilterProject` structure contains information about the logical operations, but it is not itself evaluable on data.
+Instead, it provides an `into_plan(self)` method which yields an evaluable `MfpPlan` instance.
+This means to separate out the information about what to do from the machinery to do it (for reasons that occur in the next section).
+
+The plan instances have a method `evaluate(&self, ..) -> Row` which take as an argument a `&mut Vec<Datum>`.
```

Comment:
so this wouldn't work if we wanted to e.g. filter or project things before we decoded to rows is that right?